Foundation/JSONSerialization.swift

```diff
@@ -28,6 +28,253 @@ extension JSONSerialization {
+        }
+    }
+    
+    mutating func visit<T>(_ any: T) throws {
+        if let value = any as? _JSONValue {
+            try value.visit(with: &self)
+        }
+        throw NSError(domain: NSCocoaErrorDomain, code: CocoaError.propertyListWriteInvalid.rawValue, userInfo: ["NSDebugDescription" : "Invalid value (\(any)) in JSON write"])
+    }
+    
+    mutating func visit<J:_JSONValue>(_ jsonFragment: J) throws {
+        try jsonFragment.visit(with: &self)
+    }
+}
+
+private struct JSONValidator : JSONVisitor {
+    mutating func visit(_ string: String) throws { }
+    mutating func visit<F: FloatingPoint & LosslessStringConvertible>(_ float: F) throws { }
+    mutating func visit<I: FixedWidthInteger> (_ int: I) throws { }
+    mutating func visit(_ bool: Bool) throws { }
+    mutating func visit(_ null: NSNull) throws { }
+    mutating func visit(_ decimal: Decimal) throws { }
+    mutating func visit<J: _JSONValue>(_ dictionary: [String : J]) throws {
+        for (_, value) in dictionary {
+            try value.visit(with: &self)
+        }
+    }
+    mutating func visit<J: _JSONValue>(_ array: [J]) throws {
+        for value in array {
+            try value.visit(with: &self)
+        }
+    }
+    mutating func visit(_ cocoaArray: NSArray) throws {
+        for object in cocoaArray {
+            guard let json = object as? _JSONValue else {
+                throw NSError(domain: NSCocoaErrorDomain, code: CocoaError.propertyListWriteInvalid.rawValue, userInfo: ["NSDebugDescription" : "Invalid value (\(object)) in JSON write"])
+            }
+            try json.visit(with: &self)
+        }
+    }
+    mutating func visit(_ cocoaDictionary: NSDictionary) throws {
+        var err: NSError? = nil
+        cocoaDictionary.enumerateKeysAndObjects { (key, value, stopPtr) in
+            guard let jsonKey = key as? _JSONValue else {
+                err = NSError(domain: NSCocoaErrorDomain, code: CocoaError.propertyListWriteInvalid.rawValue, userInfo: ["NSDebugDescription" : "Invalid NSDictionary key (\(key)) in JSON write"])
+                stopPtr.pointee = true
+                return
+            }
+            guard let jsonValue = value as? _JSONValue else {
+                err = NSError(domain: NSCocoaErrorDomain, code: CocoaError.propertyListWriteInvalid.rawValue, userInfo: ["NSDebugDescription" : "Invalid NSDictionary value (\(value)) in JSON write"])
+                stopPtr.pointee = true
+                return
+            }
+            do {
+                try jsonKey.visit(with: &self)
+                try jsonValue.visit(with: &self)
+            } catch where error is NSError {
+                err = (error as! NSError)
+                stopPtr.pointee = true
+            } catch {
+                fatalError()
+            }
+        }
+        if let error = err {
```

Comment:
This is because enumerateKeysAndObjects doesn't rethrow

Reply:
I don't know if this is worth a radar, but maybe it is?